controllers/propagator/aggregation.go

```diff
@@ -4,65 +4,46 @@ import (
 	"context"
 	"sort"
 
-	retry "github.com/avast/retry-go/v3"
-	appsv1 "open-cluster-management.io/multicloud-operators-subscription/pkg/apis/apps/placementrule/v1"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+	"k8s.io/apimachinery/pkg/types"
 
 	policiesv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
-	"open-cluster-management.io/governance-policy-propagator/controllers/common"
 )
 
 // calculatePerClusterStatus lists up all policies replicated from the input policy, and stores
 // their compliance states in the result list. Additionally, clusters in the failedClusters input
 // will be marked as NonCompliant in the result. The result is sorted by cluster name. An error
 // will be returned if lookup of the replicated policies fails, and the retries also fail.
 func (r *PolicyReconciler) calculatePerClusterStatus(
-	instance *policiesv1.Policy, failedClusters decisionSet,
+	instance *policiesv1.Policy, allDecisions, failedClusters decisionSet,
 ) ([]*policiesv1.CompliancePerClusterStatus, error) {
 	if instance.Spec.Disabled {
 		return nil, nil
 	}
 
-	replicatedPlcList := &policiesv1.PolicyList{}
-
-	log.V(1).Info("Getting the replicated policies")
-
-	err := retry.Do(
-		func() error {
-			return r.List(
-				context.TODO(),
-				replicatedPlcList,
-				client.MatchingLabels(LabelsForRootPolicy(instance)),
-			)
-		},
-		getRetryOptions(log.V(1), "Retrying to list the replicated policies")...,
-	)
-	if err != nil {
-		log.Info("Gave up on listing the replicated policies after too many retries")
-		r.recordWarning(instance, "Could not list the replicated policies")
-
-		return nil, err
-	}
-
-	status := make([]*policiesv1.CompliancePerClusterStatus, 0, len(replicatedPlcList.Items)+len(failedClusters))
-
-	// Update the status based on the replicated policies
-	for _, rPlc := range replicatedPlcList.Items {
-		key := appsv1.PlacementDecision{
-			ClusterName:      rPlc.GetLabels()[common.ClusterNameLabel],
-			ClusterNamespace: rPlc.GetLabels()[common.ClusterNamespaceLabel],
-		}
+	status := make([]*policiesv1.CompliancePerClusterStatus, 0, len(allDecisions))
 
-		if failed := failedClusters[key]; failed {
+	// Update the status based on the processed decisions
+	for decision := range allDecisions {
+		if failed := failedClusters[decision]; failed {
 			// Skip the replicated policies that failed to be properly replicated
 			// for now. This will be handled later.
 			continue
 		}
 
+		rPlc := &policiesv1.Policy{}
+		key := types.NamespacedName{
+			Namespace: decision.ClusterNamespace, Name: instance.Namespace + "." + instance.Name,
+		}
+
+		err := r.Get(context.TODO(), key, rPlc)
```

Comment:
So the `List` was replaced by multiple `Gets` - one for each successfully replicated policy? It's not clear to me that this will perform better. If there is an overhead to each "request" to the cache, then it feels like this would not be better.

Reply:
I replaced the `List` call not for performance reasons but due to race conditions. I found that using `allDecisions` as the authoritative list of replicated policies was more reliable since the tests would flake with missing replicated policies in the status. I believe this was due to the cache not being updated with all the new replicated policies in the time of the `List` call. This didn't show up previously because this method was called again as part of the Reconcile of the root policy status update and then it would get corrected. I since relaxed the predicates, so perhaps this is no longer an issue.

It does use less memory to get one policy at a time from the cache as long as the garbage collection happens when the single replicated policy goes out of scope.

Do you want me to try changing it back?